_drafts/cortex-m-watchpoints.md

```diff
@@ -0,0 +1,1097 @@
+        elif size == 1:
+            value = value | value << 8 | value << 16 | value << 24
+
+        # For Cortex-M4's data value matching is only supported in comparator 1
+        comp_id = 1
+        cls._write_dwt_comp_reg(DWT_COMP_BASE, comp_id, value)
+        cls._write_dwt_comp_reg(DWT_MASK_BASE, comp_id, 0)
+
+        # The range scanned can be filtered with other comparators
+        # or disabled by setting the value to the comparator being
+        # used. For this example, let's just disable:
+        #   DATAVADDR0 (bits[19:16]) = comp_id
+        #   DATAVADDR1 (bits[15:12]) = comp_id
+        #
+        # Rest of Config:
+        #   DATAVSIZE (bits[11:10]) = size >> 1
+        #   DATAVMATCH (bit[8]) = 1
+        #     Encoding is 00 (byte), 01 (half-word), 10 (word)
+        #   FUNCTION = 0x6 for a write-only watchpoint
+        datavsize = (size >> 1) << 10
+        func = (comp_id << 16) | (comp_id << 12) | (1 << 8) | datavsize | 0x6
+        cls._write_dwt_comp_reg(DWT_FUNC_BASE, comp_id, func)
+
+
+class GdbArgumentParseError(Exception):
+    pass
+
+
+class GdbArgumentParser(argparse.ArgumentParser):
+    """Wrap argparse so gdb doesn't exit when a command fails """
+
+    def exit(self, status=0, message=None):
+        if message:
+            self._print_message(message)
+        # Don't call sys.exit()
+        raise GdbArgumentParseError()
+
+
+class WatchpointExtensionCliCmd(gdb.Command):
+    """Extension of  GDB's watch for Cortex-M """
+
+    def __init__(self):
+        super(WatchpointExtensionCliCmd, self).__init__("watch_ext", gdb.COMMAND_USER)
+
+    def complete(self, text, word):
+        # We expect the argument passed to be a symbol so fallback to the
+        # internal tab-completion handler for symbols
+        return gdb.COMPLETE_SYMBOL
+
+    def invoke(self, unicode_args, from_tty):
+        # We can pass args here and use Python CLI utilities like argparse
+        # to do argument parsing
+        parser = GdbArgumentParser(
+            description="Additional Cortex-M Watchpoint Utilities"
+        )
+
+        parser.add_argument(
+            "-c", "--comp", help="The memory address or value to watch for"
+        )
+        parser.add_argument("-s", "--size", help="The size to watch")
+        parser.add_argument(
+            "--compare-value",
+            type=bool,
+            help="Watch for value written instead of address written",
+        )
+
+        try:
+            args = parser.parse_args(list(filter(None, unicode_args.split(" "))))
+        except GdbArgumentParseError:
+            return
+
+        comp = int(args.comp, 0)
+        size = int(args.size, 0)
+
+        if not args.compare_value:
+            print("Setting Address Watchpoint")
+            Dwt.set_addr_watchpoint(comp, size)
+        else:
+            print("Setting Data Value Watchpoint")
+            Dwt.set_data_value_watchpoint(comp, size)
+
+
+WatchpointExtensionCliCmd()
+```
+
```

Comment:
Maybe add a brief summary of the features implemented as a list of bullet points?

Reply:
Will watchpoints set with your script register with GDB as standard watchpoints? Can they be disabled with GDB commands or is `watchpoint_ext` required to disable them?

Reply:
Done! The script is definitely just a starting point / POC so I tried to call that out.